hc/api/models.py

```diff
@@ -14,252 +14,262 @@
+	grace = models.DurationField(default=DEFAULT_GRACE)
+	n_pings = models.IntegerField(default=0)
+	last_ping = models.DateTimeField(null=True, blank=True)
+	alert_after = models.DateTimeField(null=True, blank=True, editable=False)
+	status = models.CharField(max_length=6, choices=STATUSES, default="new")
+
+	def name_then_code(self):
+		if self.name:
+			return self.name
+
+		return str(self.code)
+
+	def url(self):
+		return settings.PING_ENDPOINT + str(self.code)
+
+	def log_url(self):
+		return settings.SITE_ROOT + reverse("hc-log", args=[self.code])
+
+	def email(self):
+		return "%s@%s" % (self.code, settings.PING_EMAIL_DOMAIN)
+
+	def send_alert(self):
+		if self.status not in ("up", "down", "fast"):
+			raise NotImplementedError("Unexpected status: %s" % self.status)
+
+		errors = []
+		for channel in self.channel_set.all():
+			error = channel.notify(self)
+			if error not in ("", "no-op"):
+				errors.append((channel, error))
+
+		return errors
+
+	def get_status(self):
+		if self.status in ("new", "paused"):
+			return self.status
+
+		now = timezone.now()
+
+		if self.last_ping + self.timeout - self.grace > now:
+			return "fast"
+		elif self.last_ping + self.timeout + self.grace > now:
+			return "up"
+		return "down"
+
+	def in_grace_period(self):
+		if self.status in ("new", "paused"):
+			return False
+
+		up_ends = self.last_ping + self.timeout
+		grace_ends = up_ends + self.grace
+		return up_ends < timezone.now() < grace_ends
+
+	def in_reverse_grace(self):
+		if self.status in ("new", "paused"):
+			return False
+
+		up_ends = self.last_ping + self.timeout
+		grace_begins = up_ends - self.grace
+		return grace_begins < timezone.now() < up_ends
+
+	def assign_all_channels(self):
+		if self.user:
+			channels = Channel.objects.filter(user=self.user)
+			self.channel_set.add(*channels)
+
+	def tags_list(self):
+		return [t.strip() for t in self.tags.split(" ") if t.strip()]
+
+	def to_dict(self):
+		pause_rel_url = reverse("hc-api-pause", args=[self.code])
+
+		result = {
+			"name": self.name,
+			"ping_url": self.url(),
+			"pause_url": settings.SITE_ROOT + pause_rel_url,
+			"tags": self.tags,
+			"timeout": int(self.timeout.total_seconds()),
+			"grace": int(self.grace.total_seconds()),
+			"n_pings": self.n_pings,
+			"status": self.get_status()
+		}
+
+		if self.last_ping:
+			result["last_ping"] = self.last_ping.isoformat()
+			result["next_ping"] = (self.last_ping + self.timeout).isoformat()
+		else:
+			result["last_ping"] = None
+			result["next_ping"] = None
+
+		return result
 
 
 class Ping(models.Model):
-    n = models.IntegerField(null=True)
-    owner = models.ForeignKey(Check)
-    created = models.DateTimeField(auto_now_add=True)
-    scheme = models.CharField(max_length=10, default="http")
-    remote_addr = models.GenericIPAddressField(blank=True, null=True)
-    method = models.CharField(max_length=10, blank=True)
-    ua = models.CharField(max_length=200, blank=True)
+	n = models.IntegerField(null=True)
```

Comment:
Why have you added extra tabs on lines in this file?

Reply:
we didn't . That was already in the file

Reply:
Ok